runtime/pendulum/src/xcm_config.rs

```diff
@@ -253,6 +300,69 @@ impl ShouldExecute for DenyReserveTransferToRelayChain {
 	}
 }
 
+match_types! {
+	pub type ParentOrParentsPlurality: impl Contains<MultiLocation> = {
+		MultiLocation { parents: 1, interior: Here } |
+		MultiLocation { parents: 1, interior: X1(Plurality { .. }) }
+	};
+}
+
+pub type Barrier = (
+	TakeWeightCredit,
+	AllowTopLevelPaidExecutionFrom<Everything>,
+	// Parent and its plurality get free execution
+	AllowUnpaidExecutionFrom<ParentOrParentsPlurality>,
+	// Expected responses are OK.
+	AllowKnownQueryResponses<PolkadotXcm>,
+	// Subscriptions for version tracking are OK.
+	AllowSubscriptionsFrom<Everything>,
+);
+
+pub struct ChargeWeightInFungiblesImplementation;
+impl ChargeWeightInFungibles<AccountId, Tokens> for ChargeWeightInFungiblesImplementation {
+	fn charge_weight_in_fungibles(
+		asset_id: CurrencyId,
+		weight: Weight,
+	) -> Result<Balance, XcmError> {
+		let amount = <WeightToFee as WeightToFeeTrait>::weight_to_fee(&weight);
+
+		// since this is calibrated (in theory) for the native of the relay
+		// we should just have a multiplier for relative "value" of that token
+		// and adjust the amount inversily proportional to the value
+		if let Some(relative_value) = RelayRelativeValue::get_relative_value(asset_id) {
+			let adjusted_amount =
+				RelativeValue::adjust_amount_by_relative_value(amount, relative_value);
+			log::info!("amount to be charged: {:?} in asset: {:?}", adjusted_amount, asset_id);
+			return Ok(adjusted_amount)
+		} else {
+			log::info!("amount to be charged: {:?} in asset: {:?}", amount, asset_id);
+			return Ok(amount)
+		}
+	}
+}
+
+/// Means for transacting the currencies of this parachain
+type Transactor = MultiCurrencyAdapter<
+	Currencies,
+	(), // We don't handle unknown assets.
+	IsNativeConcrete<CurrencyId, CurrencyIdConvert>,
+	AccountId,
+	LocationToAccountId,
+	CurrencyId,
+	CurrencyIdConvert,
+	DepositToAlternative<PendulumTreasuryAccount, Currencies, CurrencyId, AccountId, Balance>,
+>;
+
+pub type Traders = (
+	TakeFirstAssetTrader<
+		AccountId,
+		ChargeWeightInFungiblesImplementation,
+		ConvertedConcreteId<CurrencyId, Balance, CurrencyIdConvert, JustTry>,
+		Tokens,
```

Comment:
Here we should use Currencies so that we can also handle the case that someone uses our native asset for fee payments.

Reply:
Totally agree, but there is a problem, `Currencies` does not implement the necessary traits like `fungibles::Transfer`..
In any case, `Tokens` actually makes use of the type `ExistentialDeposits` which will always return `NANOUNIT`, so it can also handle the native asset. 
You can actually see that in [this modified test](https://github.com/pendulum-chain/pendulum/pull/385/files#diff-3429c5361e6214779992604d4212c994d6c327c55b74c9bcb280681ad5ddf93dR617) the treasury is getting the fee in Native.

Reply:
Oh, seems that you are right. I am almost shocked that `Currency` does not implement `fungibles`, that would be actually an important point. So how do we want to deal with a situation where someone wants to use our native asset to pay fees?

Reply:
As long we implement like [here](https://github.com/pendulum-chain/pendulum/blob/main/runtime/pendulum/src/lib.rs#L686) the existential deposit function, the tokens module will be able to give the minimum_balance and we can charge in native, as in the test.
I agree it is a bit confusing, but I cannot think of a good alternative right now.

Reply:
Seems like you are right in that the only components `TakeFirstAssetTrader` uses from this fourth generic (internally called `ConcreteAssets`) is the `AssetId`, the `Balance` and the `minimum_balance`. For the former two `Tokens` already gives the right value.

For the `minimum_balance` would just report its own value `NANOUNIT` and not the actual minimum balance we set up in pallet-balances, which is [a milliunit instead](https://github.com/pendulum-chain/pendulum/blob/fbb86dc4d27984a44abcb9ef72e89768183f238f/runtime/common/src/lib.rs#L45).

This way our `TakeFirstAssetTrader` implementation might not work correctly as it assumes a wrong minumum balance. Therefore I think that we should rewrite [our `ExistentialDeposits` function](https://github.com/pendulum-chain/pendulum/blob/fbb86dc4d27984a44abcb9ef72e89768183f238f/runtime/pendulum/src/lib.rs#L686) and explicitly return the same `EXISTENTIAL_DEPOSIT` we use for pallet-balances whenever the argument `_currency_id` is the native asset.